src/core/build_target.go

```diff
@@ -464,10 +464,12 @@ func (target *BuildTarget) AllURLs(state *BuildState) []string {
 // resolveDependencies matches up all declared dependencies to the actual build targets.
 // TODO(peterebden,tatskaari): Work out if we really want to have this and how the suite of *Dependencies functions
 //                             below should behave (preferably nicely).
-// TODO(tatskaari): Work out if we can use a channel instead of a callback.
 func (target *BuildTarget) resolveDependencies(graph *BuildGraph, callback func(*BuildTarget) error) error {
 	var g errgroup.Group
-	target.mutex.RLock()
+
+	target.mutex.Lock()
+	defer target.mutex.Unlock()
```

Comment:
I'm dubious about holding this while the waitgroup runs. Could this potentially deadlock or just slow things down a lot?